File: src/Core.Rust.DynamicValue/src/markdown.rs
use crate::{DynamicValue, EncodeError};

/// Parse a Markdown string into metadata (DynamicValue::Object) and the remaining body string.
/// Asserts strict canonical check on the frontmatter YAML.
pub fn parse_markdown(text: &str) -> Result<(DynamicValue, String), String> {
    if text.starts_with("---") && (text.len() == 3 || text.as_bytes()[3] == b'\n' || (text.as_bytes()[3] == b'\r' && text.len() > 4 && text.as_bytes()[4] == b'\n')) {
        let header_len = if text.as_bytes()[3] == b'\r' { 5 } else { 4 };

        let mut index = header_len;
        let mut close_start = None;
        let mut newline_len = 0;
        let mut close_end = 0;

        let bytes = text.as_bytes();
        while index < bytes.len() {
            let mut is_newline = false;
            let mut cur_newline_len = 0;
            let mut next_idx = index;

            if bytes[index] == b'\n' {
                is_newline = true;
                cur_newline_len = 1;
                next_idx = index + 1;
            } else if bytes[index] == b'\r' && index + 1 < bytes.len() && bytes[index + 1] == b'\n' {
                is_newline = true;
                cur_newline_len = 2;
                next_idx = index + 2;
            }

            if is_newline {
                if next_idx + 3 <= bytes.len() && &bytes[next_idx..next_idx + 3] == b"---" {
                    let tail_idx = next_idx + 3;
                    if tail_idx == bytes.len() {
                        close_start = Some(index);
                        newline_len = cur_newline_len;
                        close_end = tail_idx;
                        break;
                    } else if bytes[tail_idx] == b'\n' {
                        close_start = Some(index);
                        newline_len = cur_newline_len;
                        close_end = tail_idx + 1;
                        break;
                    } else if bytes[tail_idx] == b'\r' && tail_idx + 1 < bytes.len() && bytes[tail_idx + 1] == b'\n' {
                        close_start = Some(index);
                        newline_len = cur_newline_len;
                        close_end = tail_idx + 2;
                        break;
                    }
                }
                index = next_idx;
            } else {
                index += 1;
            }
        }

        let close_start = match close_start {
            Some(idx) => idx,
            None => return Err("Unclosed frontmatter delimiter".to_string()),
        };

        let yaml_part = &text[header_len..close_start + newline_len];
        let body_part = &text[close_end..];

        let decoded = DynamicValue::from_canonical_yaml(yaml_part)
            .map_err(|e| format!("Failed to parse frontmatter YAML: {e:?}"))?;

        match decoded {
            DynamicValue::Object(pairs) => Ok((DynamicValue::Object(pairs), body_part.to_string())),
            _ => Err("Frontmatter must be a YAML map (Object)".to_string()),
        }
    } else {
        Ok((DynamicValue::Object(Vec::new()), text.to_string()))
    }
}

/// Serialize metadata (DynamicValue::Object) and a body string into a Markdown string with frontmatter.
/// If metadata is empty (Object(vec![])), frontmatter section is omitted entirely.
pub fn serialize_markdown(metadata: &DynamicValue, body: &str) -> Result<String, EncodeError> {
    match metadata {
        DynamicValue::Object(pairs) => {
            if pairs.is_empty() {
                Ok(body.to_string())
            } else {
                let yaml = metadata.to_canonical_yaml()?;
                Ok(format!("---\n{yaml}---\n{body}"))
            }
        }
        _ => Err(EncodeError::NotXmlRepresentable),
    }
}

File: src/Core.TypeScript/dynamic-value/markdown.js
import { canonicalYaml, fromCanonicalYaml } from "./yaml";
export function parseMarkdown(text) {
    text = text ?? "";
    if (text.startsWith("---") && (text.length === 3 || text[3] === "\n" || (text[3] === "\r" && text.length > 4 && text[4] === "\n"))) {
        const headerLen = text[3] === "\r" ? 5 : 4;
        let index = headerLen;
        let closeStart = -1;
        let newlineLen = -1;
        let closeEnd = -1;
        while (index < text.length) {
            let isNewline = false;
            let curNewlineLen = 0;
            let nextIdx = index;
            if (text[index] === "\n") {
                isNewline = true;
                curNewlineLen = 1;
                nextIdx = index + 1;
            }
            else if (text[index] === "\r" && index + 1 < text.length && text[index + 1] === "\n") {
                isNewline = true;
                curNewlineLen = 2;
                nextIdx = index + 2;
            }
            if (isNewline) {
                if (nextIdx + 3 <= text.length && text.substring(nextIdx, nextIdx + 3) === "---") {
                    const tailIdx = nextIdx + 3;
                    if (tailIdx === text.length) {
                        closeStart = index;
                        newlineLen = curNewlineLen;
                        closeEnd = tailIdx;
                        break;
                    }
                    else if (text[tailIdx] === "\n") {
                        closeStart = index;
                        newlineLen = curNewlineLen;
                        closeEnd = tailIdx + 1;
                        break;
                    }
                    else if (text[tailIdx] === "\r" && tailIdx + 1 < text.length && text[tailIdx + 1] === "\n") {
                        closeStart = index;
                        newlineLen = curNewlineLen;
                        closeEnd = tailIdx + 2;
                        break;
                    }
                }
                index = nextIdx;
            }
            else {
                index++;
            }
        }
        if (closeStart === -1) {
            return { ok: false, error: "Unclosed frontmatter delimiter" };
        }
        const yamlPart = text.substring(headerLen, closeStart + newlineLen);
        const bodyPart = text.substring(closeEnd);
        const fromYamlRes = fromCanonicalYaml(yamlPart);
        if (!fromYamlRes.ok) {
            return { ok: false, error: `Failed to parse frontmatter YAML: ${fromYamlRes.error}` };
        }
        if (fromYamlRes.value.t !== "obj") {
            return { ok: false, error: "Frontmatter must be a YAML map (Object)" };
        }
        return { ok: true, metadata: fromYamlRes.value, body: bodyPart };
    }
    else {
        return { ok: true, metadata: { t: "obj", v: [] }, body: text };
    }
}
export function serializeMarkdown(metadata, body) {
    if (metadata.t !== "obj") {
        return { ok: false, error: "Metadata must be a Object" };
    }
    if (metadata.v.length === 0) {
        return { ok: true, value: body ?? "" };
    }
    const toYamlRes = canonicalYaml(metadata);
    if (!toYamlRes.ok) {
        return { ok: false, error: toYamlRes.error };
    }
    const result = `---\n${toYamlRes.value}---\n${body ?? ""}`;
    return { ok: true, value: result };
}
